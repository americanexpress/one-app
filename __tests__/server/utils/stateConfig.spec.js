/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

jest.mock('fs', () => ({ existsSync: jest.fn() }));
jest.mock('ip', () => ({ address: jest.fn() }));
jest.mock('fake/path/.dev/endpoints/index.js', () => jest.fn(), { virtual: true });
jest.mock('../../../src/server/utils/envVarAllowList', () => [
  'ONE_CLIENT_FAKE_SETTING',
  'ONE_FAKE_SETTING',
  'ONE_CLIENT_ANOTHER_SETTING',
  'ONE_CLIENT_SOME_OTHER_API_URL',
  'CLIENT_FAKE_SETTING',
]);

describe('stateConfig methods', () => {
  let setStateConfig;
  let getClientStateConfig;
  let getServerStateConfig;
  let provideStateConfig;
  let restoreModuleStateConfig;
  let backupModuleStateConfig;
  let fs;
  let ip;

  const originalEnvVars = process.env;

  const reloadMocks = () => {
    fs = require('fs');
    ip = require('ip');
    jest.spyOn(process, 'cwd').mockImplementation(() => 'fake/path/');
    ip.address.mockImplementation(() => '127.0.0.1');
    fs.existsSync.mockImplementation(() => false);
    process.env.ONE_CONFIG_ENV = 'qa';
    provideStateConfig = {
      someApiUrl: {
        client: {
          development: 'https://internet-origin-dev.example.com/some-api/v1',
          qa: 'https://internet-origin-qa.example.com/some-api/v1',
          production: 'https://internet-origin.example.com/some-api/v1',
        },
        server: {
          development: 'https://intranet-origin-dev.example.com/some-api/v1',
          qa: 'https://intranet-origin-qa.example.com/some-api/v1',
          production: 'https://intranet-origin.example.com/some-api/v1',
        },
      },
    };
  };
  describe('stateConfig', () => {
    beforeEach(() => {
      jest.resetAllMocks();
      jest.resetModules();
      reloadMocks();
    });
    afterEach(() => {
      process.env = originalEnvVars;
    });
    describe('with module config', () => {
      beforeEach(() => {
        fs.existsSync.mockImplementation(() => false);
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
          restoreModuleStateConfig,
          backupModuleStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
      });
      it('should backup and restore process module config', () => {
        setStateConfig(provideStateConfig);
        const backup = backupModuleStateConfig();
        restoreModuleStateConfig(backup);
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should throw if ONE_CONFIG_ENV is not set on parsing client', () => {
        jest.resetModules();
        delete process.env.ONE_CONFIG_ENV;
        delete provideStateConfig.someApiUrl.server;
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        const execute = () => {
          setStateConfig(provideStateConfig);
        };
        expect(execute).toThrowErrorMatchingSnapshot();
      });
      it('should throw if ONE_CONFIG_ENV is not set on parsing server', () => {
        jest.resetModules();
        delete process.env.ONE_CONFIG_ENV;
        delete provideStateConfig.someApiUrl.client;
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        const execute = () => {
          setStateConfig(provideStateConfig);
        };
        expect(execute).toThrowErrorMatchingSnapshot();
      });
      it('should set config with matching client server variables', () => {
        setStateConfig(provideStateConfig);
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should throw if client variables are supplied without server variables', () => {
        delete provideStateConfig.someApiUrl.server;
        const execute = () => setStateConfig(provideStateConfig);
        expect(execute).toThrowErrorMatchingSnapshot();
      });
      it('should throw if server variables are supplied without client variables', () => {
        delete provideStateConfig.someApiUrl.client;
        const execute = () => setStateConfig(provideStateConfig);
        expect(execute).toThrowErrorMatchingSnapshot();
      });
      it('should set config if provided a string', () => {
        provideStateConfig = {
          someApiUrl: {
            client: 'https://internet-origin-string.example.com/some-api/v1',
            server: 'https://internet-origin-string.example.com/some-api/v1',
          },
        };
        setStateConfig(provideStateConfig);
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should set config if provided a number', () => {
        provideStateConfig = {
          timeout: {
            client: 600000,
            server: 600000,
          },
        };
        setStateConfig(provideStateConfig);
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should set config if provided a boolean', () => {
        provideStateConfig = {
          enableTest: {
            client: true,
            server: true,
          },
        };
        setStateConfig(provideStateConfig);
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should not set config if provided not a string or object', () => {
        provideStateConfig = {
          someApiUrl: {
            client: null,
            server: null,
          },
        };
        setStateConfig(provideStateConfig);
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
    });
    describe('with dev endpoint config', () => {
      beforeEach(() => {
        fs.existsSync.mockImplementation(() => true);
      });
      it('should show dev endpoint supplied config', () => {
        process.env.NODE_ENV = 'development';
        // eslint-disable-next-line unicorn/import-index, import/no-unresolved
        require('fake/path/.dev/endpoints/index.js').mockImplementation(() => ({
          someOtherApiUrl: {
            devProxyPath: 'some-other-api',
            destination: 'https://intranet-origin-dev.example.com/some-other-api/v1',
          },
        }));
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
    });
    describe('with env vars', () => {
      it('should parse string undefined as js undefined', () => {
        process.env = {
          ONE_CLIENT_FAKE_SETTING: 'undefined',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should set client env var to server and client config', () => {
        process.env = {
          ONE_CLIENT_FAKE_SETTING: 'client-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should set server env var to just server config not client config', () => {
        process.env = {
          ...process.env,
          ONE_FAKE_SETTING: 'server-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should set client env var to client config but not override server config if var already exists', () => {
        process.env = {
          ...process.env,
          ONE_FAKE_SETTING: 'server-fake-setting',
          ONE_CLIENT_FAKE_SETTING: 'client-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should set client and server config from env var and merge with other variables', () => {
        process.env = {
          ...process.env,
          ONE_CLIENT_FAKE_SETTING: 'client-fake-setting',
          ONE_FAKE_SETTING: 'server-fake-setting',
          ONE_CLIENT_ANOTHER_SETTING: 'another-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should ignore env vars not prefixed with ONE', () => {
        process.env = {
          ...process.env,
          CLIENT_FAKE_SETTING: 'client-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
    });
    describe('with priorities of', () => {
      it('should prioritize dev endpoint over env', () => {
        fs.existsSync.mockImplementation(() => true);
        // eslint-disable-next-line unicorn/import-index, import/no-unresolved
        require('fake/path/.dev/endpoints/index.js').mockImplementation(() => ({
          someOtherApiUrl: {
            devProxyPath: 'some-other-api',
            destination: 'https://intranet-origin-dev.example.com/some-other-api/v1',
          },
        }));
        process.env = {
          ...process.env,
          NODE_ENV: 'development',
          ONE_CLIENT_SOME_OTHER_API_URL: 'some-other-api-url-env-var',
          ONE_CLIENT_FAKE_SETTING: 'client-fake-setting',
          ONE_FAKE_SETTING: 'server-fake-setting',
          ONE_CLIENT_ANOTHER_SETTING: 'another-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        expect(getClientStateConfig()).toMatchSnapshot();
        expect(getServerStateConfig()).toMatchSnapshot();
      });
      it('should throw if module config tries to override env var', () => {
        provideStateConfig = {
          someOtherApiUrl: {
            client: {
              development: 'https://internet-origin-dev.example.com/some-api/v1',
              qa: 'https://internet-origin-qa.example.com/some-api/v1',
              production: 'https://internet-origin.example.com/some-api/v1',
            },
            server: {
              development: 'https://internet-origin-dev.example.com/some-api/v1',
              qa: 'https://internet-origin-qa.example.com/some-api/v1',
              production: 'https://internet-origin.example.com/some-api/v1',
            },
          },
        };
        process.env = {
          ...process.env,
          NODE_ENV: 'development',
          ONE_CLIENT_SOME_OTHER_API_URL: 'some-other-api-url-env-var',
          ONE_CLIENT_FAKE_SETTING: 'client-fake-setting',
          ONE_FAKE_SETTING: 'server-fake-setting',
          ONE_CLIENT_ANOTHER_SETTING: 'another-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        const execute = () => setStateConfig(provideStateConfig);
        expect(execute).toThrowErrorMatchingSnapshot();
      });
      it('should throw if env overrides module config', () => {
        provideStateConfig = {
          someOtherApiUrl: {
            client: {
              development: 'https://internet-origin-dev.example.com/some-api/v1',
              qa: 'https://internet-origin-qa.example.com/some-api/v1',
              production: 'https://internet-origin.example.com/some-api/v1',
            },
            server: {
              development: 'https://internet-origin-dev.example.com/some-api/v1',
              qa: 'https://internet-origin-qa.example.com/some-api/v1',
              production: 'https://internet-origin.example.com/some-api/v1',
            },
          },
        };
        process.env = {
          ...process.env,
          NODE_ENV: 'development',
          ONE_CLIENT_SOME_OTHER_API_URL: 'some-other-api-url-env-var',
          ONE_CLIENT_FAKE_SETTING: 'client-fake-setting',
          ONE_FAKE_SETTING: 'server-fake-setting',
          ONE_CLIENT_ANOTHER_SETTING: 'another-fake-setting',
        };
        ({
          setStateConfig,
          getClientStateConfig,
          getServerStateConfig,
        } = require('../../../src/server/utils/stateConfig'));
        const execute = () => setStateConfig(provideStateConfig);
        expect(execute).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
