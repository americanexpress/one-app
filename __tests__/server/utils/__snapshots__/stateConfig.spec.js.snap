// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateConfig methods stateConfig with dev endpoint config should show dev endpoint supplied config 1`] = `
Object {
  "someOtherApiUrl": "http://localhost:3002/some-other-api",
}
`;

exports[`stateConfig methods stateConfig with dev endpoint config should show dev endpoint supplied config 2`] = `
Object {
  "someOtherApiUrl": "http://localhost:3002/some-other-api",
}
`;

exports[`stateConfig methods stateConfig with env vars should ignore env vars not prefixed with ONE 1`] = `Object {}`;

exports[`stateConfig methods stateConfig with env vars should ignore env vars not prefixed with ONE 2`] = `Object {}`;

exports[`stateConfig methods stateConfig with env vars should parse string undefined as js undefined 1`] = `
Object {
  "fakeSetting": undefined,
}
`;

exports[`stateConfig methods stateConfig with env vars should parse string undefined as js undefined 2`] = `
Object {
  "fakeSetting": undefined,
}
`;

exports[`stateConfig methods stateConfig with env vars should set client and server config from env var and merge with other variables 1`] = `
Object {
  "anotherSetting": "another-fake-setting",
  "fakeSetting": "client-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with env vars should set client and server config from env var and merge with other variables 2`] = `
Object {
  "anotherSetting": "another-fake-setting",
  "fakeSetting": "server-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with env vars should set client env var to client config but not override server config if var already exists 1`] = `
Object {
  "fakeSetting": "client-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with env vars should set client env var to client config but not override server config if var already exists 2`] = `
Object {
  "fakeSetting": "server-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with env vars should set client env var to server and client config 1`] = `
Object {
  "fakeSetting": "client-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with env vars should set client env var to server and client config 2`] = `
Object {
  "fakeSetting": "client-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with env vars should set server env var to just server config not client config 1`] = `Object {}`;

exports[`stateConfig methods stateConfig with env vars should set server env var to just server config not client config 2`] = `
Object {
  "fakeSetting": "server-fake-setting",
}
`;

exports[`stateConfig methods stateConfig with module config should backup and restore process module config 1`] = `
Object {
  "someApiUrl": "https://internet-origin-qa.example.com/some-api/v1",
}
`;

exports[`stateConfig methods stateConfig with module config should backup and restore process module config 2`] = `
Object {
  "someApiUrl": "https://intranet-origin-qa.example.com/some-api/v1",
}
`;

exports[`stateConfig methods stateConfig with module config should not set config if provided not a string or object 1`] = `Object {}`;

exports[`stateConfig methods stateConfig with module config should not set config if provided not a string or object 2`] = `Object {}`;

exports[`stateConfig methods stateConfig with module config should set config if provided a boolean 1`] = `
Object {
  "enableTest": true,
}
`;

exports[`stateConfig methods stateConfig with module config should set config if provided a boolean 2`] = `
Object {
  "enableTest": true,
}
`;

exports[`stateConfig methods stateConfig with module config should set config if provided a number 1`] = `
Object {
  "timeout": 600000,
}
`;

exports[`stateConfig methods stateConfig with module config should set config if provided a number 2`] = `
Object {
  "timeout": 600000,
}
`;

exports[`stateConfig methods stateConfig with module config should set config if provided a string 1`] = `
Object {
  "someApiUrl": "https://internet-origin-string.example.com/some-api/v1",
}
`;

exports[`stateConfig methods stateConfig with module config should set config if provided a string 2`] = `
Object {
  "someApiUrl": "https://internet-origin-string.example.com/some-api/v1",
}
`;

exports[`stateConfig methods stateConfig with module config should set config with matching client server variables 1`] = `
Object {
  "someApiUrl": "https://internet-origin-qa.example.com/some-api/v1",
}
`;

exports[`stateConfig methods stateConfig with module config should set config with matching client server variables 2`] = `
Object {
  "someApiUrl": "https://intranet-origin-qa.example.com/some-api/v1",
}
`;

exports[`stateConfig methods stateConfig with module config should throw if ONE_CONFIG_ENV is not set on parsing client 1`] = `"Failed to parse an object in the root module configuration due to missing ONE_CONFIG_ENV."`;

exports[`stateConfig methods stateConfig with module config should throw if ONE_CONFIG_ENV is not set on parsing server 1`] = `"Failed to parse an object in the root module configuration due to missing ONE_CONFIG_ENV."`;

exports[`stateConfig methods stateConfig with module config should throw if client variables are supplied without server variables 1`] = `
"Root module attempted to set the following non-overrideable options for the client but not the server:
  someApiUrl"
`;

exports[`stateConfig methods stateConfig with module config should throw if server variables are supplied without client variables 1`] = `
"Root module attempted to set the following non-overrideable options for the server but not the client:
  someApiUrl"
`;

exports[`stateConfig methods stateConfig with priorities of should prioritize dev endpoint over env 1`] = `
Object {
  "anotherSetting": "another-fake-setting",
  "fakeSetting": "client-fake-setting",
  "someOtherApiUrl": "http://localhost:3002/some-other-api",
}
`;

exports[`stateConfig methods stateConfig with priorities of should prioritize dev endpoint over env 2`] = `
Object {
  "anotherSetting": "another-fake-setting",
  "fakeSetting": "server-fake-setting",
  "someOtherApiUrl": "http://localhost:3002/some-other-api",
}
`;

exports[`stateConfig methods stateConfig with priorities of should throw if env overrides module config 1`] = `
"Root module attempted to set the following non-overrideable options that have been set by env var:
  someOtherApiUrl"
`;

exports[`stateConfig methods stateConfig with priorities of should throw if module config tries to override env var 1`] = `
"Root module attempted to set the following non-overrideable options that have been set by env var:
  someOtherApiUrl"
`;
